composer.json

```diff
@@ -18,6 +18,7 @@
     "phpstan/phpstan": "^1.4"
   },
   "conflict": {
+    "psr/container": "1.1.2",
```

Comment:
Why the conflict?

Reply:
Can we bump version of something in Symfony in require-dev instead of conflicting with this?

Reply:
It does not relate to Symfony but to [PHP versions](https://github.com/php-fig/container/issues/30#issuecomment-791556780). After importing `Symfony\Component\DependencyInjection\ParameterBag\ParameterBag` in `ServiceDynamicReturnTypeExtension` it implicitly imported Symfony DI exceptions and problem appeared.

The bug was already here before my changes, but code was never used and analysed, hence green checks, but if you open `Symfony\Component\DependencyInjection\Exception\ExceptionInterface` in IDE on `master` branch you'll see:

<img width="914" alt="Symfony\Component\DependencyInjection\Exception\ExceptionInterface cannot implement previously implemented interface Throwable" src="https://user-images.githubusercontent.com/600668/153561537-c46432c3-3d43-4e68-8dc5-ac4d6e9c23bf.png">

Reply:
You can put newer "psr/container" in require-dev instead.

Reply:
@ondrejmirtes can you tell me what would be the difference? 🤔  `psr/container` is required anyway:

```
$ composer why psr/container
symfony/dependency-injection  v5.4.3  requires  psr/container (^1.1.1)
symfony/service-contracts     v2.5.0  requires  psr/container (^1.1)
```

adding it to `require-dev` won't change anything since I have to add constraint `^1.1 <1.1.2` and restrict installation of 1.1.1 anyway. `conflicts` does the same but on different level of package resolution... or am I missing something?

Reply:
The difference is that I don't want the conflict for the consumers of this package. I want them to be able to use it because the problematic part might not even be executed in their use-case.

Reply:
OK, I'll change it.

-----

But IMHO using `require-dev` here is only dirty hack. In fact, e.g. `symfony/dependency-injection` is used in the actual code (not in dev processes only) and it should be strict requirement, installable on consumers' side even without explicit requirement in their `composer.json`. Right now you only assume that those packages will be installed. [Composer Require Checker](https://github.com/maglnet/ComposerRequireChecker) reports that 56 unknown symbols were found.

Also, you support PHP7.1 in Github Actions' matrix while there's (dev) requirement for `symfony/http-foundation ^5.1` which requires PHP >=7.2.5... Looking at the [changelog](https://github.com/symfony/http-foundation/blob/5.4/CHANGELOG.md#500) and usages of `Symfony\Component\HttpFoundation` namespace in this package there's probably no problem if consumer has `symfony/http-foundation` v4, but in general it's not good that Symfony extension uses Symfony components without requiring them explicitly. In fact, this works in Github checks only because of [platform overriding](https://github.com/phpstan/phpstan-symfony/blob/master/composer.json#L40), but on actual PHP7.1 it's not installable:

```
/app # composer require --dev phpunit/phpunit:^7.5.20 -w
./composer.json has been updated
Running composer update phpunit/phpunit --with-dependencies
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - symfony/http-foundation[v5.1.0-BETA1, ..., v5.1.0-RC1] require php ^7.2.5 -> your php version (7.1.33) does not satisfy that requirement.
    - symfony/http-foundation[v5.1.0-RC2, ..., 5.4.x-dev] require php >=7.2.5 -> your php version (7.1.33) does not satisfy that requirement.
    - Root composer.json requires symfony/http-foundation ^5.1 -> satisfiable by symfony/http-foundation[v5.1.0-BETA1, ..., 5.4.x-dev].
```

Also usage of Symfony packages v6 is not restricted on consumer's side (because you require `^4.2 || ^5.0` only for development) and I'm not sure if it'll work properly 🤔 

Sorry for this rant 😅 I don't want to blame you for anything, I am just wondering if this is proper approach in terms of QA for this package. Feel free to ignore it in terms of this MR, but I really would like to know your opinion and reasoning behind current architecture.